src/config/decoding.rs

```diff
@@ -0,0 +1,141 @@
+pub struct DecodingsConfig {
+    pub decoding: Vec<DecodingConfig>,
+}
+
+impl<'de> Deserialize<'de> for DecodingsConfig {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct DecodingsConfig {
+            decoding: DecodingsConfigValue,
+        }
+
+        #[derive(Deserialize)]
+        #[serde(untagged)]
+        enum DecodingsConfigValue {
+            Single(DecodingConfig),
+            Multiple(Vec<DecodingConfig>),
+        }
+
+        let config = DecodingsConfig::deserialize(deserializer)?;
+
+        Ok(Self {
+            decoding: match config.decoding {
+                DecodingsConfigValue::Single(config) => vec![config],
+                DecodingsConfigValue::Multiple(configs) => configs,
+            },
+        })
+    }
+}
+
+impl From<Vec<DecodingConfig>> for DecodingsConfig {
+    fn from(configs: Vec<DecodingConfig>) -> Self {
+        Self { decoding: configs }
+    }
+}
+
+impl IntoIterator for DecodingsConfig {
+    type Item = DecodingConfig;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.decoding.into_iter()
+    }
+}
+
+#[derive(Serialize, Debug, Eq, PartialEq, Clone)]
+pub struct DecodingConfig(pub(crate) Decoding);
+
+impl DecodingConfig {
+    pub fn decode(self, event: Event) -> Event {
+        let decoder = Into::<Decoder>::into(self);
+
+        decoder.decode(event)
+    }
+}
+
+impl<'de> Deserialize<'de> for DecodingConfig {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct StringOrStruct;
+
+        impl<'de> Visitor<'de> for StringOrStruct {
+            type Value = DecodingConfig;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("string or map")
+            }
+
+            fn visit_str<E>(self, value: &str) -> std::result::Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(DecodingConfig(Decoding::deserialize(
+                    value.into_deserializer(),
+                )?))
+            }
+
+            fn visit_map<M>(self, map: M) -> std::result::Result<Self::Value, M::Error>
+            where
+                M: MapAccess<'de>,
+            {
+                Deserialize::deserialize(de::value::MapAccessDeserializer::new(map))
+            }
+        }
+
+        deserializer.deserialize_any(StringOrStruct)
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum Decoding {
+    Utf8,
+    Json,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use indoc::indoc;
+
+    #[test]
+    fn config_decodings_single() {
+        let config: DecodingsConfig = toml::from_str(indoc! {r#"
+            decoding = "json"
```

Comment:
Do you think you'll stick with just `decoding`? At the sink level, we moved what use to be `encoding` down into `encoding.codec` in order put `only_fields` and `except` fields alongside. We could consider making this `decoding.codec` just for parity and in-case we want to add additional fields alongside.

I could even see this being:

```
decoding.codec = "" (or maybe `format`)
decoding.framing =""
decoding.compression = ""
```

Have you thought about compression yet? Would you think that would be a separate field or a chained encoding? Notably it seems like it would need to be applied _before_ the framing.

Reply:
We'll definitely want be able to define options under the `decoding` key, I just decided to defer implementing them until I know which ones we need.

> Have you thought about compression yet? Would you think that would be a separate field or a chained encoding? Notably it seems like it would need to be applied before the framing.

That's a very good question, and I haven't given it any thought yet. I'd lean to taking it out of scope for now

Reply:
The configuration now supports all of these forms:

```toml
codec = "json"
```

```toml
codec = { type = "json", target_field = "foo" }
```

```toml
codec = ["syslog", { type = "json", target_field = "foo" }]
```

Implementations for `Framer`/`Codec` register themselves in the `inventory` with a short-hand name and an arbitrary amount of additional option fields via `typetag::serde`.

Reply:
Nice! I like that.